LIP-38: VEBO-7002 (#127)
lidofinance/lido-improvement-proposals: LIPS/lip-38.md

---
lip: 38
title: "VEBO-7002 — Withdrawal-Based Exits and Active Rebalancing"
status: WIP
author: Raman Siamionau (@F4ever)
discussions-to: <Create a new thread on https://research.lido.fi/ and drop the link here>
created: 2026-06-26
---

## Simple Summary

This proposal redesigns the Validators Exit Bus Oracle (VEBO) around EIP-7002 execution-layer withdrawal requests and introduces **Active Rebalancing** — a mechanism that lets the protocol redistribute stake across Node Operators without waiting for organic inflows and outflows.

1. **VEBO-7002** — instead of relying on Node-Operator-supplied pre-signed exit messages and the Ejector as the primary exit tool, VEBO uses EIP-7002 partial withdrawal requests (PWRs) and full withdrawal requests (FWRs) as the primary exit mechanism. This enables precise, partial stake exits from `0x02` validators, reduces unproductive stake time, shortens and sharpens withdrawal finalization, and simplifies the protocol by removing several legacy components.

> 2. **Active Rebalancing** — a controlled, rate-limited way for the protocol to forcibly move stake toward a target distribution by requesting additional withdrawals from over-target Community Module v2 (CMv2) operators, reusing the same withdrawal-based flow introduced by VEBO-7002.

Reviewer: ```suggestion
2. **Active Rebalancing** — a controlled, rate-limited way for the protocol to forcibly move stake toward a target distribution by requesting additional withdrawals from over-target Curated Module v2 (CMv2) operators, reusing the same withdrawal-based flow introduced by VEBO-7002.
```

**On-chain**, we add a **FIFO queue** to the VEBO that stores EIP-7002-based withdrawal requests (partial and full) submitted by the Oracle report, together with a **permissionless handle** that anyone can call to trigger and execute queued requests one by one via the EIP-7002 withdrawal-request contract. A `setPartialWithdrawals` switch lets the protocol turn partial withdrawals off, reverting to FWR-only operation (that could be fulfilled through a Validator Ejector) under extreme EIP-7002 fee conditions.

**Off-chain**, we rework the exit-ordering logic so it selects not only which validators to exit but also the withdrawal amount per validator. This adds support for **partial withdrawal requests** — preferring PWRs against `0x02` validators and falling back to FWRs where PWRs are not applicable — and for **Active Rebalancing**: additional, rate-limited exits against CMv2 operators whose `currentStake` exceeds their `targetStake`, moving the module toward its target distribution.

This redesign also removes components made redundant by the new flow: late-exit penalties and proofs to trigger full withdrawal requests.

## Motivation

### Limitations of the current VEBO

> The current VEBO depends on Node Operators uploading pre-signed validator exit messages, and on the Ejector software to broadcast those exits. This design has structural limitations:

Reviewer: ```suggestion
The current VEBO depends on Node Operators uploading pre-signed validator exit messages and the Ejector software to broadcast those exits. This design has structural limitations:
```

- **Full exits only.** The flow exclusively supports full validator exits. Partial withdrawals — increasingly the right tool after EIP-7251 (MaxEB) and `0x02` credentials — are not currently available.

> - **Limited upload capacity.** Only a fraction of pre-signed exit messages can be uploaded per Node Operator, to bound the blast radius if an operator's signing infrastructure is compromised. This caps how aggressively exits can be requested.

Reviewer: ```suggestion
- **Limited upload capacity.** Only a fraction of pre-signed exit messages can be uploaded per Node Operator, to bound the blast radius if an operator's signing infrastructure is compromised. This caps how aggressively exits can be requested and what validators can be exited using Ejector software.
```

- **Operational overhead.** Every Node Operator must run and secure dedicated Ejector software, and the protocol must verify that they do.
- **Coarse exit optimization.** Because only specific pre-signed keys can be requested, optimizing exit order (e.g. to minimize sweep delay or to target precise amounts of stake) is difficult.

### Why withdrawal-based exits

EIP-7002 lets the protocol trigger withdrawals from the execution layer using validators' `0x02` withdrawal credentials, without holding pre-signed messages. Building VEBO around PWRs and FWRs:

- Allows **precise stake exit** from `0x02` validators in CMv2 using partial withdrawals, leaving the validator active.

> - **Eliminates unproductive stake time** — ETH keeps earning until the moment a withdrawal is actually triggered, instead of sitting idle behind a queued full exit.

Reviewer: ```suggestion
- **Eliminates unproductive stake time** — ETH keeps earning rewards until the moment a withdrawal is actually processed, instead of sitting idle behind a queued full exit and waiting for the sweep to reach exited validator.
```

## Specification

### Overview

1. **Off-chain Oracle computes exit demand in three sequential phases.**
    - Phase 1 — cover regular **withdrawal-queue demand**.
    - Phase 2 — issue **forced validator exits**.
    - Phase 3 — add **active rebalancing** within CMv2. Phase 3 is optional and can be switched off (leaving only Phases 1–2).
2. **The Oracle reports the ordered requests on-chain, where they are appended to a single FIFO queue in the VEBO.**
3. **Anyone can execute items in the queue.** A permissionless handle pops queued items one by one and submits them to the EIP-7002 predeploy as partial (PWR) or full (FWR) withdrawal requests, paying the per-request fee.

A **`setPartialWithdrawals(bool)` switch** can turn partial withdrawals off, making VEBO emit FWRs only; `ExitRequested` events are emitted for all FWRs so a simplified Ejector can fulfill them via voluntary exits without EIP-7002 fees.

### Rationale

#### Single FIFO queue

Withdrawal requests (both PWRs and FWRs) are stored in a single FIFO queue and processed in submission order. We chose this model because:

- **Simplicity.** A single queue is simple to iterate on-chain, with no need to manage concurrency.

> - **Permissionless safety.** VEBO fully controls ordering and concurrency, which protects against permissionless interference and mitigates FWR-replay DDoS.

Reviewer: Add a link to the report with the DDoS finding

The trade-offs are:

- **No skipping.** Items must be executed in order; an item that is no longer valid still must be executed to process order further. We deliberately do not add an item-skipping mechanism: EIP-7002 requests are cheap, so executing a now-invalid request wastes only a negligible fee, which is not worth the extra on-chain complexity.

> - **No execution-order choice.** Because order is fixed, there is no room for execution-time optimization (e.g. minimizing FWR sweep delay). This is acceptable because we expect to use mostly PWRs, so no sweep optimization is needed.

Reviewer: ```suggestion
- **No execution-order choice.** Because order is fixed, there is no room for execution-time optimization (e.g. minimizing FWR sweep delay). This is acceptable because we expect to use mostly PWRs for CMv2, so no sweep optimization is needed. Instances of CSM (0x01 and 0x02) do not support PWRs due to the FIFO stake distribution mechanism and the requirement for the key to be fully bonded.
```

##### Shared iterator mechanics

All three phases reuse the same machinery. Two governance-tunable parameters drive it, both stored in the `OracleDaemonConfig` smart contract so they can be changed without an Oracle release:

- **`EXIT_ITERATION_CHUNK`** (default **32 ETH**) — the fixed unit of demand the iterator allocates per step.
- **`MIN_PARTIAL_WITHDRAWAL`** (default **8 ETH**) — the minimum withdrawable balance a validator must have above the effective-balance floor to serve a PWR, and the lower bound of any single PWR.

**Iteration granularity.** The iterator walks exit demand in fixed `EXIT_ITERATION_CHUNK` (32 ETH) steps: each iteration allocates one chunk of demand to the top-ranked operator (and one of its validators), then re-ranks. This keeps the ordering balanced across operators and maps naturally onto withdrawal sizing — a chunk becomes a PWR of up to `EXIT_ITERATION_CHUNK` from a `0x02` validator, or an FWR.

> **PWR vs. FWR threshold.** A validator can only serve a PWR if it has enough balance above the effective-balance floor to give up. If **no** validator has more than `MIN_PARTIAL_WITHDRAWAL` (8 ETH) of withdrawable balance above the floor, a **full withdrawal request** is used against the top-ranked validator instead. Otherwise a partial withdrawal is taken from that validator, bounded to the `[MIN_PARTIAL_WITHDRAWAL, EXIT_ITERATION_CHUNK]` (8–32 ETH) range per validator.

Reviewer: From here, it is not clear that we can make a PWR of more than 32 ETH if several ordering iterations end up with the same validator.

**Per-report exit limit.** The iterator stops accumulating requests once the report's cumulative exited balance (`max_current_exit_balance`) reaches the per-report exit limit. This limit is shared across all three phases: Phase 1 draws against it first, and Phases 2–3 consume whatever remains.

##### Phase 1 — Cover withdrawal-queue demand

Phase 1 covers the current **withdrawal-queue (WQ) demand**, issuing PWR-first instead of full exits. It walks that demand in `EXIT_ITERATION_CHUNK` steps, ranking validator at each step by, in order:

1. **Forced-exit deviation** — operators over a force-exit limit first, by how far `currentValidators − targetValidators` exceeds the threshold.
2. **Soft-exit deviation** — same idea for soft-exit limits, after forced exits.
3. **Module share-rate deviation** — modules above their `priority_exit_share_threshold` (relative to total protocol balance) get elevated priority.

> 4. **Target-stake deviation** — operators furthest **above** their target stake exit first, where `targetStake = moduleStake × operatorWeight / moduleTotalWeight`.

Reviewer: This is only for CMv2 tho

6. **Biggest validator first** — among the operator's validators, prefer the one with the **largest balance**. This maximizes the balance a single PWR can extract while leaving the validator active, favoring partial over full withdrawals.
7. **Lowest validator index** — within an operator, ascending validator index.

##### Phase 2 — Forced validator exits

After WQ demand is covered, the iterator satisfies any **forced-exit obligations**.

1. **Forced-exit deviation** — operators over a force-exit limit first, by how far `currentValidators − targetValidators` exceeds the threshold.
2. **Lowest validator index** — within an operator, ascending validator index.

##### Phase 3 — Active rebalancing

This phase applies **only to the CMv2 staking module** and adds exit demand against operators drifting above their target stake, pushing the module back toward its target distribution.

> **Configuration.** All Phase 4 parameters live as keys in the existing **`OracleDaemonConfig`** key-value store.

Reviewer: What's Phase 4?

| `OracleDaemonConfig` key                | Type        | Meaning                                                                                     |
|-----------------------------------------|-------------|---------------------------------------------------------------------------------------------|

> | `ACTIVE_REBALANCING_ENABLED`            | bool        | Global on/off for Phase 3.                                                                  |

Reviewer: Would it be configurable by the DAO?

| `ACTIVE_REBALANCING_RATE_LIMIT`         | uint (ETH)  | Max stake exitable for rebalancing per VEBO report.                                         |
| `ACTIVE_REBALANCING_OPERATOR_EXCESS_BP` | uint (bp)   | Operator-level trigger: `currentStake` over `targetStake` as a share of `targetStake`.      |
| `ACTIVE_REBALANCING_GRACE_PERIOD`       | uint (days) | Grace period before a new operator can affect the module's stake balance used to rebalance. |

**Trigger.** Phase 3 runs only if `ACTIVE_REBALANCING_ENABLED` is set. 

A **new-operator grace period** applies: operators created and whose first key activated less than `ACTIVE_REBALANCING_GRACE_PERIOD` days ago cannot *affect* rebalancing calculations, but remain eligible to *receive* redistributed deposits via the normal deposit allocation.

**Bounds.** The total rebalancing stake added to the report is capped at `ACTIVE_REBALANCING_RATE_LIMIT`.

> **Ordering.** When it runs, the iterator ranks over-target CMv2 operators by, in order:

Reviewer: ACTIVE_REBALANCING_OPERATOR_EXCESS_BP should not be an ordering thing but a trigger to launch Phase 3. If there is an operator (or several operators) above this limit Phase 3 is activated.

We also miss a threshold for global imbalance. When none of the operators is above ACTIVE_REBALANCING_OPERATOR_EXCESS_BP but the overall imbalance in the module is above some threshold

```
VEBO-7002 (FIFO queue)  ->  TriggerableWithdrawalsGateway (TWG)  ->  WithdrawalVault  ->  EIP-7002 predeploy
  report submission                 rate limits, fee refund             encoding         withdrawal request
```

##### Flow 1 — Report submission

The Oracle calls `submitReportData` with a packed report. The report keeps the existing encoding — a `dataFormat` selector plus a `data` blob of fixed-width records packed contiguously — but introduces a **new `dataFormat` version** whose record carries the withdrawal **amount**. Today's record is 64 bytes; the new record appends an 8-byte `amount`:

```
/// MSB <----------------------------------------------------------------------------------- LSB
/// |  3 bytes   |   5 bytes    |    8 bytes     |      48 bytes       |      8 bytes          |
/// |  moduleId  |  nodeOpId    | validatorIndex |   validatorPubkey   |  amount (gwei) *new*  |
```

`amount` (uint64, gwei) — **new**; `0` = full withdrawal (FWR), `> 0` = partial withdrawal (PWR).

On submission the contract decodes each record into an `ExitRequest` struct and appends it to the FIFO queue. 

> When `setPartialWithdrawals` is off, the report is still in the new format but every record MUST carry `amount == 0` (FWR-only); a non-zero amount is rejected.

Reviewer: Is a full report rejected, or just the invalid request?

##### Flow 2 — Execute withdrawal requests

Anyone calls the permissionless `processExitRequests(count)`, forwarding the per-request EIP-7002 fee. It pops the next `count` requests from the FIFO in order and pushes each down the path above: **TWG** applies the exit limits and refunds excess fee, then the **WithdrawalVault** encodes the request and submits it to the EIP-7002 contract. A request rejected at the predeploy is dropped and must be re-submitted by a later Oracle report.

```solidity
interface IValidatorsExitBusOracle7002 {
    /// A single queued withdrawal request.
    /// amountGwei == 0  -> full withdrawal request (FWR)
    /// amountGwei  > 0  -> partial withdrawal request (PWR)
    struct ExitRequest {
        uint256 moduleId;

>         uint256 nodeOperatorId;

Reviewer: can be uint128 for effective packing

        bytes   pubkey;
        uint64  amountGwei;
    }

    /// Flow 1: Oracle submits the ordered, packed report; requests are appended to the FIFO queue.
    function submitReportData(bytes calldata report, uint256 contractVersion) external;

    /// Flow 2: permissionless — pop and deliver the next `count` queued requests to the
    /// EIP-7002 withdrawal-request path. Caller forwards the per-request fee.
    function processExitRequests(uint256 count) external payable;

    /// Number of requests waiting in the FIFO queue.

>     function unprocessedRequestsCount() external view returns (uint256);

Reviewer: What about the view method for the queue?

    /// FWR-only fallback switch (privileged).
    function setPartialWithdrawals(bool enabled) external;
    function partialWithdrawalsEnabled() external view returns (bool);

    /// Emitted for every request (validator ejector can use this events 
    /// to fulfill exits when partial withdrawals are turned off).
    event ExitRequested(
        uint256 indexed moduleId,
        uint256 indexed nodeOperatorId,
        bytes pubkey,
        uint64 amountGwei
    );
}
```

> ##### Scope — contracts to change

Reviewer: Also modules - remove late exit logic and penalties

|-----------------------------------------|-------------|---------|-------------------------------------------------------------------------------------------|
| `EXIT_ITERATION_CHUNK`                  | uint (ETH)  | 32      | Fixed unit of demand the iterator allocates per step.                                     |
| `MIN_PARTIAL_WITHDRAWAL`                | uint (ETH)  | 8       | Min withdrawable balance above the floor for a validator to serve a PWR; PWR lower bound. |
| `ACTIVE_REBALANCING_ENABLED`            | bool        | false   | Global on/off for Phase 3.                                                                |
| `ACTIVE_REBALANCING_RATE_LIMIT`         | uint (ETH)  | —       | Max stake exitable for rebalancing per VEBO report.                                       |
| `ACTIVE_REBALANCING_OPERATOR_EXCESS_BP` | uint (bp)   | —       | Operator-level trigger: `currentStake` over `targetStake` as a share of `targetStake`.    |
| `ACTIVE_REBALANCING_GRACE_PERIOD`       | uint (days) | —       | Grace period before a new operator can affect the module's stake balance.                 |

#### EasyTrack factories

TODO

> The EasyTrack factories that interact with VEBO (e.g. the VEBO-bypass exit flow) are affected by the new report format and withdrawal-based flow. They can either be **removed** (if no longer needed after the CMv1/SDVT sunset) or **adapted** to the new VEBO-7002 report format.

Reviewer: I suggest removing them

- **Permissionless interference / DDoS.** The single FIFO queue keeps concurrency entirely under VEBO control and avoids FWR-replay attack surface.
- **PWR/FWR concurrency on one validator.** The consensus layer rejects a full withdrawal request while the validator still has an unprocessed partial withdrawal request. So if VEBO issues a PWR and then an FWR for the same validator, the FWR is rejected by the CL and must be re-requested later. To avoid this waste, **VEBO MUST NOT issue an FWR for a validator that still has an unprocessed PWR** — it should wait until the PWR is processed (or only ever escalate a validator from PWR to FWR across separate reports once the PWR has cleared).
- **Balance-denominated TWG limit.** Changing the TWG rate limiter to bound extracted balance (instead of request count) caps how much stake can leave per frame, so a stream of large partial withdrawals cannot exceed the intended ETH-denominated limit. Note this makes FWRs the expensive request against the limit: an FWR consumes the validator's full max effective balance (up to 2048 ETH for `0x02`) from the frame budget, so a burst of FWRs can exhaust the TWG limit far faster than PWRs.

> - **EIP-7002 fee risk.** Organic-usage modeling shows fee escalation is a minimal credible risk at realistic batch sizes; turning `setPartialWithdrawals` off (FWR-only operation) provides a safe path under sustained extreme fees.

Reviewer: Add a link to the research by Max

Author: Done

---
lip: 38

> title: "VEBO-7002 — Withdrawal-Based Exits and Active Rebalancing"

Reviewer: ```suggestion
title: "VEBO-7002 — Execution Layer initiated Exits and Active Rebalancing"
```

status: WIP

> author: Raman Siamionau (@F4ever)

Reviewer: I would appreciate my name here as well since the original design doc was a joint effort

Author: Added

discussions-to: <Create a new thread on https://research.lido.fi/ and drop the link here>
created: 2026-06-26
updated: 2026-07-24
---

## Simple Summary

> This proposal redesigns the Validators Exit Bus Oracle (VEBO) around EIP-7002 execution-layer withdrawal requests and introduces **Active Rebalancing** — a mechanism that lets the protocol redistribute stake across Node Operators without waiting for organic inflows and outflows.

Reviewer: ```suggestion
This proposal redesigns the Validators Exit Bus Oracle (VEBO) around **EIP-7002 execution-layer withdrawal requests** and introduces **Active Rebalancing** — a mechanism that lets the protocol redistribute stake across Node Operators without waiting for organic inflows and outflows.
```

## Specification

### Overview

1. **Off-chain Oracle computes exit demand in three sequential phases.**
    - Phase 1 — cover regular **withdrawal-queue demand**.
    - Phase 2 — issue **forced validator exits**.
    - Phase 3 — add **active rebalancing** within CMv2. Phase 3 is optional and can be switched off (leaving only Phases 1–2).
2. **The Oracle reports the ordered withdrawal request intentions on-chain, where they are appended to a single FIFO queue in the VEBO.**
3. **Anyone can execute queued intentions.** A permissionless handle pops them one by one and submits each as an actual partial (PWR) or full (FWR) withdrawal request to the EIP-7002 predeploy, paying the per-request fee.

> A **`setPartialWithdrawals(bool)` switch** can turn partial withdrawals off, making VEBO emit FWRs only; `ExitRequested` events are emitted for all FWRs so a simplified Ejector can fulfill them via voluntary exits without EIP-7002 fees.

Reviewer: ```suggestion
A **`enablePartialWithdrawals(bool)` switch** can turn partial withdrawals off, making VEBO emit FWRs only; `ExitRequested` events are emitted for all FWRs so a simplified Ejector can fulfill them via voluntary exits without EIP-7002 fees.
```

Not sure about the exact name, but the current one is not really straightforward

Author: Splitted into two methods
```
enablePartialWithdrawals() / disablePartialWithdrawals()
```

Reviewer 2: Do we really need two methods?

```
 enablePartialWithdrawals()
 disablePartialWithdrawals()
```

I think just one `enablePartialWithdrawals(bool)` with parameter should be fine.

- **Simplicity.** A single queue is simple to iterate on-chain, with no need to manage concurrency.
- **Permissionless safety.** VEBO fully controls ordering and concurrency, which protects against permissionless interference and mitigates [FWR-replay DDoS](https://github.com/lidofinance/lido-improvement-proposals/blob/develop/LIPS/lip-30.md#61-attack-vector-vulnerability-from-uncontrolled-creation-of-twrs).

The trade-offs are:

- **No skipping.** Intentions must be executed in order; an intention that is no longer valid still must be executed to process order further. We deliberately do not add a skipping mechanism: EIP-7002 requests are cheap, so executing a now-invalid intention wastes only a negligible fee, which is not worth the extra on-chain complexity.
- **No execution-order choice.** Because order is fixed, there is no room for execution-time optimization (e.g. minimizing FWR sweep delay). This is acceptable because we expect to use mostly PWRs for CMv2, so no sweep optimization is needed. Instances of CSM (0x01 and 0x02) do not support PWRs due to the FIFO stake distribution mechanism and the requirement for the key to be fully bonded.

> - **PWR/FWR concurrency.** PWRs and FWRs share one queue and contend for the same EIP-7002 throughput; this concurrency will be mitigated within VEBO (in how it orders and batches requests). Concurrency matters at the CL level for a given validator: a PWR cannot be executed while the validator is already exiting (an FWR/voluntary exit is pending), and an FWR cannot be executed while the validator still has an unprocessed PWR pending — either combination causes the later request to be rejected on the CL and re-requested later (see [Shared iterator mechanics](#shared-iterator-mechanics) and [Security Considerations](#security-considerations)).

Reviewer: I think this is not a trade-off. Or at least not fully a trade-off. FIFO guarantees that VEBO has a full control over the order and can prevent concurrency

Author: 100%. This advance described above on line 71

#### Deprecating the Ejector as the primary tool

> The Ejector cannot process PWRs, the new system expects a low FWR volume, and historical mainnet data shows that EIP-7002 fee spikes are short-lived and resolve before VEBO could react. With a declining validator count and no planned reduction in EIP-7002 throughput, PWR pricing should stay favorable. The Ejector is therefore demoted to a **fallback** role, used only when partial withdrawals are turned off via `setPartialWithdrawals(bool)` (FWR-only operation).

Reviewer: ```suggestion
The Ejector cannot process PWRs, the new system expects a low FWR volume, and historical mainnet data shows that EIP-7002 fee spikes are short-lived and resolve before VEBO could react. With a declining validator count and no planned reduction in EIP-7002 throughput, EIP-7002 requests pricing should stay favorable. The Ejector is therefore demoted to a **fallback** role, used only when partial withdrawals are turned off via `setPartialWithdrawals(bool)` (FWR-only operation).
```

Author: Fixed

    - enforces the exit limits (TWG rate limit on extracted balance);
    - enforces the configurable rebalancing bounds (rate limit, thresholds);
    - holds partial withdrawal requests off/on switch (full withdrawal requests only mode);

#### Off-chain Oracle: ordering and rebalancing logic

Per report, the Oracle produces the ordered list of withdrawal request intentions. It runs the exit-order iterator ([`exit_order_iterator.py`](https://github.com/lidofinance/lido-oracle/blob/f352b8552b717ebf8ecaf3bd0f5cf6785909ec17/src/services/exit_order_iterator.py#L106)) — generalized from "pick the next validators to fully exit" to "pick the next validators and the amount (PWR/FWR) to withdraw" — across **three sequential phases**, each with its own demand source.

##### Shared iterator mechanics

All three phases reuse the same machinery. Two governance-tunable parameters drive it, both stored in the `OracleDaemonConfig` smart contract so they can be changed without an Oracle release:

- **`EXIT_ITERATION_CHUNK`** (default **32 ETH**) — the fixed unit of demand the iterator allocates per step.

> - **`MIN_PARTIAL_WITHDRAWAL`** (default **8 ETH**) — the minimum withdrawable balance a validator must have above the min-effective-balance floor to serve a PWR, and the lower bound of any single PWR.

Reviewer: Can be 2 ETH IMO

**Iteration granularity.** The iterator walks exit demand in fixed `EXIT_ITERATION_CHUNK` (32 ETH) steps: each iteration allocates one chunk of demand to the top-ranked operator (and one of its validators), then re-ranks. This keeps the ordering balanced across operators and maps naturally onto withdrawal sizing — a chunk becomes a PWR of up to `EXIT_ITERATION_CHUNK` from a `0x02` validator, or an FWR.

> **PWR vs. FWR threshold.** A validator can only serve a PWR if it has enough balance above the effective-balance floor to give up. If **no** validator has more than `MIN_PARTIAL_WITHDRAWAL` (8 ETH) of withdrawable balance above the floor, a **full withdrawal request** is used against the top-ranked validator instead. Otherwise a partial withdrawal is taken from that validator, bounded to the `[MIN_PARTIAL_WITHDRAWAL, EXIT_ITERATION_CHUNK]` (8–32 ETH) range per validator per iteration step. Across iteration steps within the same report, the ranking criterion can allocate multiple chunks to the same validator. These are unified into a single PWR intention for that validator.

Reviewer: ```suggestion
**PWR vs. FWR threshold.** A validator can only serve a PWR if it has enough balance above the effective-balance floor to give up. If **no** validator for the top-ranked Node Operator has more than `MIN_PARTIAL_WITHDRAWAL` (8 ETH) of withdrawable balance above the floor, a **full withdrawal request** is used against the top-ranked validator instead. Otherwise a partial withdrawal is taken from that validator, bounded to the `[MIN_PARTIAL_WITHDRAWAL, EXIT_ITERATION_CHUNK]` (8–32 ETH) range per validator per iteration step. Across iteration steps within the same report, the ranking criterion can allocate multiple chunks to the same validator. These are unified into a single PWR intention for that validator.
```

6. **Biggest validator first** — among the operator's validators, prefer the one with the **largest balance**. This maximizes the balance a single PWR can extract while leaving the validator active, favoring partial over full withdrawals.
7. **Lowest validator index** — within an operator, ascending validator index.

##### Phase 2 — Forced validator exits

After WQ demand is covered, the iterator satisfies any **forced-exit obligations**.

1. **Forced-exit deviation** — operators over a force-exit limit first, by how far `currentValidators − targetValidators` exceeds the threshold.
2. **Lowest validator index** — within an operator, ascending validator index.

##### Phase 3 — Active rebalancing

This phase applies **only to the CMv2 staking module** and adds exit demand against operators drifting above their target stake, pushing the module back toward its target distribution.

**Data model.** Phase 3 (and Phase 1 criterion 4) uses the following definitions:

> - **`operatorWeight` / `moduleTotalWeight`** — CMv2 operator weights introduced in [LIP-35](lip-35.md), read on-chain via `getOperatorWeights`. `moduleTotalWeight` is the sum of weights over operators participating in the calculation (see the grace period below).

Reviewer: gracePeriod operators should also be excluded from `moduleTotalWeight`

| `OracleDaemonConfig` key                | Type        | Meaning                                                                                                                             |
|-----------------------------------------|-------------|-------------------------------------------------------------------------------------------------------------------------------------|
| `ACTIVE_REBALANCING_ENABLED`            | bool        | Global on/off for Phase 3.                                                                                                          |
| `ACTIVE_REBALANCING_RATE_LIMIT`         | uint (ETH)  | Max stake exitable for rebalancing per VEBO report.                                                                                 |
| `ACTIVE_REBALANCING_OPERATOR_EXCESS_BP` | uint (bp)   | Operator-level trigger: `currentStake` over `targetStake` as a share of `targetStake`.                                              |
| `ACTIVE_REBALANCING_MODULE_EXCESS_BP`   | uint (bp)   | Module-level trigger: the sum of over-target stake across all operators as a share of total module stake.                           |

> | `ACTIVE_REBALANCING_GRACE_PERIOD`       | uint (days) | Days since first key activation before an operator counts toward `moduleStake`/`moduleTotalWeight` and participates in rebalancing. |

Reviewer: ```suggestion
| `ACTIVE_REBALANCING_GRACE_PERIOD`       | uint (days) | Days since the first key deposit before an operator counts toward `moduleStake`/`moduleTotalWeight` and participates in rebalancing. |
```

**Trigger.** Phase 3 requires `ACTIVE_REBALANCING_ENABLED` to be set, and runs when **one or more** of the following conditions holds:

1. **Operator-level** — at least one operator's `currentStake` exceeds `targetStake × (1 + ACTIVE_REBALANCING_OPERATOR_EXCESS_BP)`;
2. **Module-level** — the sum of stake above target stake across all operators within the module exceeds a defined percentage of the total module stake.
   `Σ max(0, currentStake − targetStake) > moduleStake × ACTIVE_REBALANCING_MODULE_EXCESS_BP`
   (summed over operators outside the grace period).

While neither condition holds, Phase 3 stays dormant, so small transient drifts don't generate exit churn.

> A **new-operator grace period** applies: an operator whose first key activated less than `ACTIVE_REBALANCING_GRACE_PERIOD` days ago is **excluded from the rebalancing calculation entirely** — its balance does not count toward `moduleStake`, its weight does not count toward `moduleTotalWeight`, and it is neither ranked nor targeted by rebalancing exits. Grace-period operators remain eligible to receive deposits via the normal deposit allocation.

Reviewer: ```suggestion
A **new-operator grace period** applies: an operator whose first key was deposited less than `ACTIVE_REBALANCING_GRACE_PERIOD` days ago is **excluded from the rebalancing calculation entirely** — its balance does not count toward `moduleStake`, its weight does not count toward `moduleTotalWeight`, and it is neither ranked nor targeted by rebalancing exits. Grace-period operators remain eligible to receive deposits via the normal deposit allocation.
```

Author: fixed

```
VEBO-7002 (FIFO queue)  ->  TriggerableWithdrawalsGateway (TWG)  ->  WithdrawalVault  ->  EIP-7002 predeploy
  report submission                 rate limits, fee refund             encoding          withdrawal request
```

##### Flow 1 — Report submission

The Oracle calls `submitReportData` with a packed report. The report keeps the existing encoding — a `dataFormat` selector plus a `data` blob of fixed-width records packed contiguously — but introduces a **new `dataFormat` version** whose record carries the withdrawal **amount**. The current record (`DATA_FORMAT_LIST_WITH_KEY_INDEX = 2`, introduced in [LIP-35](lip-35.md)) is 72 bytes; the new record appends an 8-byte `amount`:

```
/// MSB <--------------------------------------------------------------------------------------------------- LSB
/// |  3 bytes   |   5 bytes    |    8 bytes     |   8 bytes  |      48 bytes       |      8 bytes          |
/// |  moduleId  |  nodeOpId    | validatorIndex |  keyIndex  |   validatorPubkey   |  amount (gwei) *new*  |
```

`amount` (uint64, gwei) — **new**; `0` = full withdrawal (FWR), `> 0` = partial withdrawal (PWR).

> `keyIndex` is **retained** from format 2: it lets the contract check against the staking module's key registry that the reported pubkey is a real registered key and determine its withdrawal-credentials type (`0x01`/`0x02`) on-chain — this is what backs the `0x02` validation of PWRs at submission.

Reviewer: I guess you can get WC from moduleId

Author: Since it's not new field after srv3 - removed description from lip

On submission the contract decodes each record into an `ExitRequest` struct — a withdrawal request intention — and appends it to the FIFO queue.

When `setPartialWithdrawals` is off, the report is still in the new format but every record MUST carry `amount == 0` (FWR-only); if any record carries a non-zero amount, the protocol reverts the entire `submitReportData` call.

##### Flow 2 — Execute withdrawal requests

Anyone calls the permissionless `processExitRequests(count)`, forwarding the per-request EIP-7002 fee. It pops the next `count` intentions from the FIFO in order and pushes each down the path above: **TWG** applies the exit limits and refunds excess fee, then the **WithdrawalVault** encodes an actual withdrawal request and submits it to the EIP-7002 contract.

> **Cost model.** EIP-7002 fees for all VEBO-path requests are paid by the `processExitRequests` caller (excess is refunded via the TWG). This release **deliberately removes** the previous charge-back of forced-exit fees to Node Operator bond (CSM's `_fulfillExitObligations`, driven by the `onValidatorExitTriggered` notification): with the notification hooks gone, forced-exit fees are no longer billed to the misbehaving operator and are borne by the executor instead. The trade this buys: forced exits no longer depend on operator compliance at all — the protocol withdraws the stake itself via EIP-7002.

Reviewer: `_fulfillExitObligations` will still be present but for the different purpose (strikes penalties)

        uint64  nodeOperatorId; // matches the 5-byte nodeOpId field width in the packed report record
        uint64  amountGwei;     // moduleId + nodeOperatorId + amountGwei pack into a single storage slot
        bytes   pubkey;         // dynamic type placed last so it doesn't break packing of the fields above
    }

    /// Flow 1: Oracle submits the ordered, packed report; intentions are appended to the FIFO queue.
    function submitReportData(bytes calldata report, uint256 contractVersion) external;

    /// Flow 2: permissionless — pop the next `count` queued intentions and execute each as an
    /// actual withdrawal request down the EIP-7002 path. Caller forwards the per-request fee.
    function processExitRequests(uint256 count) external payable;

    /// Number of intentions waiting in the FIFO queue.
    function unprocessedRequestsCount() external view returns (uint256);

    /// Returns the queued intention at FIFO offset `index` from the head of the queue
    /// (`index == 0` is the next intention `processExitRequests` will execute), letting
    /// permissionless executors and monitoring tools inspect the queue before calling it.

>     function getExitRequest(uint256 index) external view returns (ExitRequest memory);

Reviewer: I suggest making it `getExitRequest(uint256 startIndex, uint256 count)`

Author: Hm, it will allow front-running the execution of WRs by sending a single WR to process. In the case of a huge number of WRs, it can delay execution.

| `ValidatorsExitBusOracle` (→ VEBO-7002) | New report `dataFormat` with `amount`; FIFO queue storage; permissionless `processExitRequests` handle; `setPartialWithdrawals` switch.                                                                                                                                                          |
| `TriggerableWithdrawalsGateway` (TWG)   | Add `triggerWithdrawals` (partial and full, used by VEBO-7002); **retain `triggerFullWithdrawals`** unchanged for existing consumers (CSM `Ejector`); drop `_notifyStakingModules`; rate limiter changed to bound **extracted balance** instead of request count, shared by both entry points.   |
| `StakingRouter`                         | Remove `onValidatorExitTriggered` (no longer called by the TWG) and `reportValidatorExitDelay` (late-exit-penalty accounting); revoke the two orphaned role grants.                                                                                                                              |

> | Staking Modules (CSM)                   | Remove the module-side late-exit logic and penalty accounting: the `onValidatorExitTriggered` implementations, exit-delay-penalty bookkeeping fed by `reportValidatorExitDelay`, and (CSM) the `_fulfillExitObligations` bond charge-back; see [Staking Modules](#staking-modules-nor-sdvt-csm). |

Reviewer: ```suggestion
| Staking Modules (CSM, CMv2)                   | Remove the module-side late-exit logic and penalty accounting: the `onValidatorExitTriggered` implementations, exit-delay-penalty bookkeeping fed by `reportValidatorExitDelay`, and (CSM) the `_fulfillExitObligations` bond charge-back; see [Staking Modules](#staking-modules-nor-sdvt-csm). |
```

Author: fixed

**Bounds.** The total rebalancing stake added to the report is capped at `ACTIVE_REBALANCING_RATE_LIMIT` and by the remaining shared per-report exit limit. Rebalancing reduces excess stake; it does not guarantee that the withdrawn ETH is redeposited or that it reaches a specific under-target operator. That depends on future withdrawal demand, available buffer ETH, and the normal deposit-allocation rules.

**Ordering.** When it runs, the iterator ranks over-target CMv2 operators by, in order:

1. **Target-stake deviation** — node operators rank by the deviation `currentStake − targetStake`; the operator furthest **above** target ranks first.
2. **Already-requested validators** — validators with PWR demand already allocated **within this report** rank first, concentrating exited balance into as few validators as possible.

> 3. **Biggest validator first** — prefer the largest-balance validator so exits stay partial and keep the validator active.

Reviewer 2: Could you please clarify.
So I have 3 validators each one with  2048 ETH.

Then we use PWD, as result 

1. 1000
2. 2048
3. 2048

After couple days when all pending request completed we run rebalancing again.

1. 1000
2. 1500 
3. 2048

Look like as result we will end with many validators with balance less then MaxEB

- **Execute withdrawal requests** — a permissionless handle pops queued intentions and pushes each down the request path to the EIP-7002 predeploy, where it becomes an actual withdrawal request:

```
VEBO-7002 (FIFO queue)  ->  TriggerableWithdrawalsGateway (TWG)  ->  WithdrawalVault  ->  EIP-7002 predeploy
  report submission                 rate limits, fee refund             encoding          withdrawal request
```

##### Flow 1 — Report submission

The Oracle calls `submitReportData` with the existing `ReportData` structure: a `dataFormat` selector and a `data` blob of fixed-width records packed together. It introduces a **new `dataFormat` version** whose record carries the withdrawal **amount**. The current record (`DATA_FORMAT_LIST_WITH_KEY_INDEX = 2`, introduced in [LIP-35](lip-35.md)) is 72 bytes; the new record appends an 8-byte `amount`:

```
/// MSB <--------------------------------------------------------------------------------------------------- LSB
/// |  3 bytes   |   5 bytes    |    8 bytes     |   8 bytes  |      48 bytes       |      8 bytes          |

> /// |  moduleId  |  nodeOpId    | validatorIndex |  keyIndex  |   validatorPubkey   |  amount (gwei) *new*  |

Reviewer 2: Do we need validatorIndex?

        uint32  moduleId;       // matches the 3-byte moduleId field width in the packed report record
        uint64  nodeOperatorId; // matches the 5-byte nodeOpId field width in the packed report record
        uint64  amount;         // gwei; 0 = full withdrawal (FWR), > 0 = partial withdrawal (PWR).
        uint8   wcType;         // validator withdrawal-credentials type: 0x01 or 0x02
        bytes   pubkey;         // dynamic type placed last, consistent with ExitRequest
    }

    /// New single path for partial and full withdrawal requests (used by VEBO-7002).
    /// Applies the extracted-balance rate limit, forwards each request to the WithdrawalVault,
    /// and refunds any unused EIP-7002 fee to `refundRecipient`.
    function triggerWithdrawals(
        WithdrawalRequest[] calldata requests,
        address refundRecipient
    ) external payable;

    /// Retained: full-withdrawal-only entry point with the existing external
    /// signature, kept for backward compatibility with deployed consumers
    /// (CSM Ejector). Since CSM is a 0x01 module, each request should consume 
    /// 32 ETH of the limit quota.
    function triggerFullWithdrawals(

>         IStakingRouter.ValidatorExitData[] calldata validatorsData,

Reviewer 2: Since we are going to remove the dependency on `StakingRouter` by sunsetting the `onValidatorExitTriggered` method, we can move the `ValidatorExitData` struct to `TriggerableWithdrawalsGateway`.
